Start to port Wasmtime to the new wasi-io API with resources.

sunfishcode opened the pull request:
This imports the new APIs posted in WebAssembly/wasi-io#46.

<!--
Please make sure you include the following information:

- If this work has been discussed elsewhere, please include a link to that
  conversation. If it was discussed in an issue, just mention "issue #...".

- Explain why this change is needed. If the details are in an issue already,
  this can be brief.

Our development process is documented in the Wasmtime book:
https://docs.wasmtime.dev/contributing-development-process.html

Please ensure all communication follows the code of conduct:
https://github.com/bytecodealliance/wasmtime/blob/main/CODE_OF_CONDUCT.md
-->

alexcrichton commented on crates/wasi/src/preview2/host/filesystem.rs:
```diff
@@ -15,16 +19,26 @@ impl From<TableError> for types::Error {
+            .ctx()
+            .preopens
+            .iter()
+            .map(|(fd, name)| (Resource::new_borrow(fd.rep()), name.clone()))
```
This is I believe going to return an error at runtime because borrows can't be returned, they can only be taken.

Two ways I can think of to handle this are:

* Don't store preopens as `Resource<Descriptor>`. Instead insert them into the table every time `get-directories` is called.
* Alternatively on `HostDescriptor::drop` don't actually delete preopens from the table. This will "leak" them in the table but that's ok since it's a static set that doesn't grow.

sunfishcode replied:
I made it insert them into the table every tine `get-directories` is called.

alexcrichton commented on crates/wasi/src/preview2/host/filesystem/sync.rs:
```diff
@@ -316,97 +398,167 @@ impl<T: async_filesystem::Host> sync_filesystem::Host for T {
-            async_filesystem::Host::is_same_object(self, a, b).await
+            async_filesystem::HostDescriptor::is_same_object(
+                self,
+                Resource::new_borrow(a.rep()),
+                Resource::new_borrow(b.rep()),
```
Bah this is unfortunate (as are all the other instances of this in this file). If I understand this correctly, this is because the bindings generated. have two distinct `Descriptor` types meaning that they're not compatible?

If so I can try to add a feature to `bindgen!` so resources can be specified as opposed to always-generated

sunfishcode replied:
Yes, exactly. All the `sync` adapter code has to jump through these hoops for that reason.

alexcrichton commented on crates/wasi/src/preview2/poll.rs:
```diff
@@ -121,28 +122,81 @@ impl<T: WasiView> poll::Host for T {
+            }
+        }
+
+        struct PollOne<'a> {
+            elem: PollableFuture<'a>,
+        }
+        impl<'a> Future for PollOne<'a> {
+            type Output = Result<()>;
+
+            fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+                match self.elem.as_mut().poll(cx) {
+                    Poll::Ready(Ok(())) => Poll::Ready(Ok(())),
+                    Poll::Ready(Err(e)) => Poll::Ready(Err(e.context("poll_one"))),
+                    Poll::Pending => Poll::Pending,
+                }
+            }
+        }
+
+        Ok(PollOne {
+            elem: closure_future,
+        }
+        .await?)
+    }
```
I think this can all be skipped in favor fo `closure_future.await`?

`poll-list` is a bit funky since there's more than one future, but here there's just one so I think it should be ok to do a normal `.await`

sunfishcode replied:
Done.

sunfishcode commented on crates/wasi-preview1-component-adapter/src/descriptors.rs:
```diff
@@ -24,96 +23,85 @@ pub enum Descriptor {
 pub struct Streams {
     /// The input stream, if present.
-    pub input: Cell<Option<InputStream>>,
+    pub input: UnsafeCell<Option<InputStream>>,
```
I tried that, but `OnceCell` pulls in Rust error handling which breaks the adapter build.

sunfishcode commented on crates/wasi-preview1-component-adapter/src/lib.rs:
```diff
@@ -95,6 +92,28 @@ impl<T, E> TrappingUnwrap<T> for Result<T, E> {
     }
 }
 
+trait TrappingUnwrapRef<T> {
```
Oh, good idea.

pchickey commented on crates/wasi-preview1-component-adapter/src/lib.rs:
```diff
@@ -444,7 +463,7 @@ pub unsafe extern "C" fn fd_allocate(fd: Fd, offset: Filesize, len: Filesize) ->
 #[no_mangle]
 pub unsafe extern "C" fn fd_close(fd: Fd) -> Errno {
-    State::with(|state| {
+    State::with_mut(|state| {
```
I had gotten rid of State::with_mut in https://github.com/bytecodealliance/wasmtime/pull/6973, is there a reason this PR brought it back that I haven't spotted, or is this just a git history artifact?

sunfishcode replied:
`get_input_stream()` now needs to return a `&Descriptor` instead of a `Descriptor` aka `u32`. That lifetime constraint appears to make it incompatible with holding the descriptor table in a `RefCell`. This `State::with_mut` and the current `descriptors(&self) -> &mut Descriptors` code are me searching for a way to describe what this code is doing. I don't think I've found the right answer yet.

sunfishcode replied:
I've now removed `State::with_mut`.

sunfishcode commented on crates/wasi-preview1-component-adapter/src/lib.rs:
```diff
@@ -2396,7 +2434,7 @@ impl State {
 
     #[cold]
-    fn new() -> &'static RefCell<State> {
+    fn new() -> &'static UnsafeCell<State> {
```
I've now removed the `with_mut` and this outer `UnsafeCell`.

I think the scariest part now is the `fn descriptors_mut(&self) -> &mut Descriptors`, because I couldn't get `RefCell`/`RefMut` to work now that `get_input_stream` needs to return a `&Descriptor` instead of a `u32` that it can just copy out. I'll poke around at this to see if I can figure out a better option.

sunfishcode commented on crates/wasi/src/preview2/host/logging.rs:
```diff
@@ -0,0 +1,13 @@
+use crate::preview2::{bindings::logging::logging, WasiView};
+
+impl<T: WasiView> logging::Host for T {
```
Yeah, I can do that. It's convenient for now, because with stderr being a host stream that gets buffered, it's extra important for debugging to have some kind of debug facilty that can dump text directly to the host stderr. But when this PR is closer to landing I can split it out.

alexcrichton replied:
Hm that seems like an issue with our stderr design then as well, that should definitely be suitable for debugging. We don't have any writes though that we don't push out via tokio though right? Is the issue that it's a slightly deferred write rather than immediate?

sunfishcode replied:
I've now removed the logging code.

sunfishcode commented on crates/wasi-preview1-component-adapter/src/lib.rs:
```diff
@@ -2363,28 +2393,36 @@ enum AllocationState {
+        }
+    }
+
+    fn with_mut(f: impl FnOnce(&mut State) -> Result<(), Errno>) -> Errno {
```
I've now fully removed this.

sunfishcode commented:
Ok, I've now fixed all the `get_mut`/`with_mut` issues, with a solution that I think works fairly well, which ends up implementing some amount of `RefCell` because `RefCell` itself pulls in static initializers.

And I've now updated the HTTP code to use handles for polling and streams.

Things that still need to be done:
 - Make stdio functions callable multiple times somehow, or add some kind of debug output API.
 - Figure out why cargo-vet doesn't like wit-bindgen 0.12

alexcrichton commented:
If you run `cargo vet` locally does it fix the vet issues? (I think I'm listed there as "trust anything published by alex")

sunfishcode commented:
I get the same error locally:
```
$ cargo vet --locked
Vetting Failed!

5 unvetted dependencies:
  wit-bindgen:0.12.0 missing ["safe-to-deploy"]
  wit-bindgen-core:0.12.0 missing ["safe-to-deploy"]
  wit-bindgen-rust:0.12.0 missing ["safe-to-deploy"]
  wit-bindgen-rust-lib:0.12.0 missing ["safe-to-deploy"]
  wit-bindgen-rust-macro:0.12.0 missing ["safe-to-deploy"]
````

alexcrichton commented:
Ah if you remove `--locked` it should fill in the relevant bits to commit here

sunfishcode commented:
If I remove `--locked`, it fails in a different way.
```
$ cargo vet 
ERROR   × There are some issues with your policy.audit-as-crates-io entries

Error:   × Some non-crates.io-fetched packages match published crates.io versions
  │   wasmtime-wmemcheck:14.0.0
  help: Add a `policy.*.audit-as-crates-io` entry for them

```
I get this on the `main` branch too.

alexcrichton commented:
Hm are you using 0.8.0 of `cargo vet`? I updated the wit-bindgen dep to 0.12.0 locally, ran `cargo vet`, and it succeeded, but I don't know what that specific error is

sunfishcode commented:
Yes, this is cargo-vet 0.8.0. I've now added what appear to be the needed lines for cargo-vet to be happy with wasmtime-wmemcheck, so I've now updated it for wit-bindgen too, and it seems ok now.

alexcrichton reviewed (approved):
Ok I've re-reviewed everything here and it all looks good-to-go to me. There's some follow-up cleanups I think but nothing that needs to block this PR 👍

pchickey commented on crates/test-programs/wasi-tests/src/bin/poll_oneoff_files.rs:
```diff
@@ -261,7 +261,7 @@ unsafe fn test_poll_oneoff(dir_fd: wasi::Fd) {
     test_empty_poll();
     test_fd_readwrite_valid_fd(dir_fd);
-    test_fd_readwrite_invalid_fd();
+    //test_fd_readwrite_invalid_fd(); // FIXME: This traps now, rather than failing
```
Is this still broken? Just want to note it for a follow-up so we don't forget to go back and fix it.

sunfishcode replied:
Earlier I had assumed it was failing due to the issue with calling `get_stdout` multiple times, but it turns out to have been a bug in the poll_oneoff adapter. I've now pushed a patch fixing it.

pchickey commented on crates/wasi-preview1-component-adapter/src/macros.rs:
```diff
@@ -3,23 +3,11 @@
-    }};
+pub fn println(message: &[u8]) {
+    // FIXME: We need some way to print a message.
+    let _ = message;
```
We can run `get_stderr().blocking_write_and_flush()` here now, correct?

sunfishcode replied:
Yep, this is now done.

pchickey commented on crates/wasi-preview1-component-adapter/src/macros.rs:
```diff
@@ -28,14 +16,16 @@ macro_rules! eprintln {
+        // We use `str` instead of `str_nl` because we're calling the logging
+        // API which expects lines.
+        let message = byte_array_literals::str!($arg);
+        $crate::macros::println(&message);
```
We can revert the line-oriented changes and move this back to stderr, right?

sunfishcode replied:
Yes, I've now reverted them.

pchickey reviewed (approved):
Looks good. Some minor stuff I stumbled on, and we'd like to land https://github.com/bytecodealliance/wasmtime/pull/7091 first this afternoon, which will create some minor conflicts.
